test/jdk/java/awt/ScrollPane/ScrollPaneLimitation.java

```diff
@@ -0,0 +1,148 @@
+import java.awt.Robot;
+import java.awt.ScrollPane;
+
+import java.awt.event.InputEvent;
+import java.awt.event.MouseEvent;
+import java.awt.event.MouseAdapter;
+
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+public class ScrollPaneLimitation {
+    static final int SCROLL_POS = 50000;
+    public static Component child = null;
+    static final CountDownLatch go = new CountDownLatch(1);
+    public Frame frame;
+    volatile Point point;
+    ScrollPane pane;
+
+    public static void main(String[] args) throws Exception {
+        ScrollPaneLimitation scrollTest = new ScrollPaneLimitation();
+        scrollTest.init();
+        scrollTest.start();
+    }
+
+    public void init() throws Exception {
+        EventQueue.invokeAndWait(() -> {
+            frame = new Frame("Scroll Pane Limitation");
+            frame.setLayout(new BorderLayout());
+            pane = new ScrollPane();
+            frame.add(pane);
+            child = new MyPanel();
+            child.addMouseListener(new MouseAdapter() {
+                public void mousePressed(MouseEvent e) {
+                    if (e.getID() == MouseEvent.MOUSE_PRESSED
+                            && e.getSource() == ScrollPaneLimitation.child
+                            && e.getY() > SCROLL_POS) {
+                        go.countDown();
+                    }
+                }
+            });
+            pane.add(child);
+            frame.setSize(200, 200);
+            frame.pack();
+            frame.setLocationRelativeTo(null);
+            frame.setAlwaysOnTop(true);
+            frame.setVisible(true);
+            pane.doLayout();
+        });
+    }
+
+    public void start() throws Exception {
+        try {
+            Robot robot = new Robot();
+            robot.waitForIdle();
+            robot.delay(1000);
+
+            EventQueue.invokeAndWait(() -> {
+                Point p = child.getLocation();
+                System.out.println("Child's initial location " + p);
+                System.out.println("Pane's insets " + pane.getInsets());
+                pane.setScrollPosition(0, SCROLL_POS);
+                p = pane.getScrollPosition();
+                System.out.println("Scroll pos = " + p);
+                if (p.y != SCROLL_POS) {
+                    throw new RuntimeException("wrong scroll position");
+                }
+                p = child.getLocation();
+                System.out.println("Child pos = " + p);
+                if (p.y != -SCROLL_POS) {
+                    if (child.isLightweight()) {
+                        // If it is lightweight it will always have (0, 0) location.
+                        // Check location of its parent - it is Panel and it should
+                        // be at (inset left, inset top + position)
+                        Container cp = child.getParent();
+                        p = cp.getLocation();
+                        if (p.y != -SCROLL_POS) {
```

Comment:
```suggestion
                        p = cp.getLocation();
                        System.out.println("Child's parent pos = " + p);
                        if (p.y != -SCROLL_POS) {
```
Just to make it clear that this branch is taken. Otherwise the test ends with the message:
```text
Child pos = java.awt.Point[x=0,y=0]
```
which is confusing. Yet no exception is thrown.

Reply:
Child location is expected to be 0,0 as explained in the test, whereas parents position is -500, hence no exception is thrown.

Reply:
Yes, this is adding the suggested print will resolve all the confusion because you will see:

```text
...
Scroll pos = java.awt.Point[x=0,y=50000]
Child pos = java.awt.Point[x=0,y=0]
Child's parent pos = java.awt.Point[x=0,y=-50000]
```

Isn't it clearer?

Reply:
Yeah, can do it.

Reply:
Thank you!